### PR TITLE
TRUNK-6202 Replace Hibernate Criteria API with JPA for HibernateCohortDAO

### PR DESCRIPTION
**Pull request by k4pran**
## Description of what I changed

This task is to migrate the deprecated hibernate criteria to jpa criteria for the `HibernateCohortDAO` class.

Additionally, I've incorporated several new tests and confirmed their compatibility with the existing Hibernate Criteria implementation.

I also created an additional XML to add some tests for result order validation, trying to change some of the Cohort names in the current xml failed a lot of tests so seemed more risky than creating a new xml for this purpose.

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6202 

## Checklist: I completed these to help reviewers :)

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java [hunk 1]
```diff
@@ -106,47 +111,59 @@ public Cohort deleteCohort(Cohort cohort) throws DAOException {
...
+		List<Predicate> predicates = new ArrayList<>();
+
 		if (!includeVoided) {
-			criteria.add(Restrictions.eq(VOIDED, false));
+			predicates.add(cb.isFalse(root.get(VOIDED)));
 		}
-		
-		return (List<Cohort>) criteria.list();
+
+		query.where(predicates.toArray(new Predicate[0]));
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java:148]: Wouldn't it be enough to just do? `query.where(cb.isFalse(root.get(VOIDED)))`

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java [hunk 2]
```diff
@@ -50,52 +51,56 @@ public void setSessionFactory(SessionFactory sessionFactory) {
...
+		predicates.add(cb.equal(membershipJoin.get("patientId"), patientId));
+
 		if (!includeVoided) {
-			criteria.add(Restrictions.eq(VOIDED, includeVoided));
+			predicates.add(cb.equal(root.get(VOIDED), includeVoided));
 		}
-		return criteria.list();
-	}
 
+		query.distinct(true).where(predicates.toArray(new Predicate[0]));
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java:85]: How about using this? `predicates.toArray(new Predicate[]{})`

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java [hunk 3]
```diff
@@ -157,22 +174,34 @@ public Cohort saveCohort(Cohort cohort) throws DAOException {
...
+			predicates.add(cb.or(endDateIsNull, endDateIsGreater));
 		}
+
 		if (!includeVoided) {
-			criteria.add(Restrictions.eq(VOIDED, false));
+			predicates.add(cb.isFalse(root.get(VOIDED)));
 		}
-		return criteria.list();
+
+		query.where(predicates.toArray(new Predicate[0]));
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java:202]: Same here `new Predicate[]{}`

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java [hunk 4]
```diff
@@ -50,52 +51,56 @@ public void setSessionFactory(SessionFactory sessionFactory) {
...
-	@SuppressWarnings("unchecked")
 	public List<Cohort> getCohortsContainingPatientId(Integer patientId, boolean includeVoided,
-	                                                  Date asOfDate) throws DAOException {
-		Disjunction orEndDate = Restrictions.disjunction();
-		orEndDate.add(Restrictions.isNull("m.endDate"));
-		orEndDate.add(Restrictions.gt("m.endDate", asOfDate));
+													  Date asOfDate) throws DAOException {
+		Session session = sessionFactory.getCurrentSession();
+		CriteriaBuilder cb = session.getCriteriaBuilder();
+		CriteriaQuery<Cohort> query = cb.createQuery(Cohort.class);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java:65]: For the sake of consistency, between CriteriaBuilder and CriteriaQuery, why not any of these?
1) cb and cq
2) builder and query
3) criteriaBuilder and criteriaQuery

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java (hunk 1 above)
Comment (k4pran) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java]: Yes that's more concise, updated.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java (hunk 2 above)
Comment (k4pran) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java]: I wasn't familiar with that syntax but I agree it is more readable, updated.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java (hunk 3 above)
Comment (k4pran) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java]: Done

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java (hunk 4 above)
Comment (k4pran) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java]: Updated to use option 1 (cb and cq)

**Pull request merged**